docs: reorder `skaffold debug` docs (#6894)
GoogleContainerTools/skaffold: docs/content/en/docs/workflows/debug.md

aliases: [/docs/how-tos/debug]
---

> Skaffold lets you debug your application running on a local or remote Kubernetes cluster, almost exactly how you'd do it if the code were running locally on your machine. Skaffold does this by detecting the runtime of your project and transparently configuring the containers in pods for debugging against that framework.

Reviewer: ```suggestion
Skaffold lets you debug your application running on a local or remote Kubernetes cluster, almost exactly how you would do it if the code were running locally on your machine. Skaffold does this by detecting the runtime of your project and transparently configuring the containers in pods for debugging against that framework.
```

  - Python 3.5+ (runtime ID: `python`) using `debugpy` (Debug Adapter Protocol) or `pydevd`
  - .NET Core (runtime ID: `netcore`) using `vsdbg`

> Skaffold can usually detect the correct language runtime if present. However if you're having difficulty getting it running for your application then checkout the [Supported Language Runtimes]({{< relref "#supported-language-runtimes">}}) section for the exact heuristics that skaffold uses and you can modify your application accordingly.

Reviewer: ```suggestion
Skaffold can usually detect the correct language runtime if present. However if you encounter difficulties then checkout the [Supported Language Runtimes]({{< relref "#supported-language-runtimes">}}) section for the exact heuristics that skaffold uses and you can modify your application accordingly.
```

## (Recommended) Debugging using Cloud Code

The easiest way to debug using Skaffold is by using the [Cloud Code IDE extension]({{< relref "../install/#managed-ide">}}) for Visual Studio Code, JetBrains IDEs, and Cloud Shell. Cloud Code will automatically configure container images for debugging so you can debug Kubernetes services just like how you would debug a local service in your IDE. 

For more information, see the corresponding documentation for [VS Code](https://cloud.google.com/code/docs/vscode/debug), [IntelliJ](https://cloud.google.com/code/docs/intellij/kubernetes-debugging) and [Cloud Shell](https://cloud.google.com/code/docs/shell/debug).

## Detailed debugger configuration and setup

This section describes how Skaffold configures the container image for debugging for the supported language runtimes, and how to setup with certain IDEs.

Note that many debuggers may require additional information for the location of source files.
We are looking for ways to identify this information and to pass it back if found.

> {{% tabs %}}

Reviewer: Neat idea

```
{{< /alert >}}

{{% /tab %}}
{{% /tabs %}}

## Supported Language Runtimes

This section describes how `debug` recognizes the language runtime used in a

> container image for specific language runtimes.

Reviewer: Rather than splitting this section out, why don't we integrate it into the previous tabs.  Users of most of the runtimes will have to look into this sections if only to ensure they set the right environment variables.

Author: I wanted to separate this out because this section is independently relevant to Cloud Code. So when debug isn't working in Cloud Code for a language runtime they can link this section directly. Having it merged with the previous section would present additional info that IDE users don't necessarily need to know immediately.

  `PYTHONHASHSEED`, `PYTHONDONTWRITEBYTECODE`, or
- the container command-line invokes `python`, `python2`, or `python3`.

{{% /tab %}}
{{% tab ".NETCORE" %}}

#### .NET Core (runtime: `dotnet`)

.NET Core application are recognized by:
- the presence of a standard .NET environment variable:
  `ASPNETCORE_URLS`, `DOTNET_RUNNING_IN_CONTAINER`,
  `DOTNET_SYSTEM_GLOBALIZATION_INVARIANT`, or
- the container command-line invokes the [dotnet](https://github.com/dotnet/sdk) cli

Furthermore, your app must be built with the `--configuration Debug` options to disable optimizations.

{{% /tab %}}
{{% /tabs %}}

## How It works

`skaffold debug` acts like `skaffold dev`, but it configures containers in pods
for debugging as required for each container's runtime technology.
The associated debugging ports are exposed and labelled so that they can be port-forwarded to the
local machine.
One notable difference from `skaffold dev` is that `debug` disables image rebuilding and
syncing as it leads to users accidentally terminating debugging sessions by saving file changes.
These behaviours can be re-enabled with the `--auto-build`, `--auto-deploy`, and `--auto-sync`

> flags.

Reviewer: I think this should work:

```suggestion
flags, or triggering a [devloop iteration]({{< relref "/docs/design/api#control-api" >}}) using the Skaffold API.
```
